Allow using server name as a variable in join messages, take 2

unilock opened the pull request:
I closed #9 by accident because I do not know how this website works :)

This merge changes `onConnect()` to listen to `ServerConnectedEvent` rather than `LoginEvent`, allowing it to handle both initial connections to the proxy ("\<player\> joined the game") as well as switches between backend servers ("\<player\> moved from \<previous_server\> to \<server\>").
`onConnect()` now sends either the `join` message or `switch` message depending on whether `previousServer` has a value - `switch` if it does, `join` if it doesn't.
This deprecates `onServerConnect()`.
This also allows the use of the name of the server being joined as a variable in `join` messages, as well as the use of the name of the server being switched *from* in `switch` messages.
The default config has been edited as such.

I feel like it might be possible to call `sendMessage()` only once, outside of the conditionals - maybe by tweaking `sendMessage()` itself?
Ahh, too tired to think at the moment... let's call it an exercise for the reader.

unilock committed:
Consolidate onConnect() and onServerConnect() using ServerConnectedEvent
+ Allow using server name as a variable in join messages
(+ Allow using both current and previous server names as variables in server switch messages)

Signed-off-by: unilock <unilock@fennet.rentals>

unilock committed:
Update default config

Signed-off-by: unilock <unilock@fennet.rentals>

BBaoVanC reviewed (approved):
Looks good, sorry for late review: I had this buried under hundreds of old GitHub notifications

@confor can you look at this? (Haven't compiled and tested myself but it looks right)

confor commented:
i'm too busy studying but i'll try to set up a network this weekend. i remember there was some bug in velocity while testing in an offline server [1] that prevented me from using `ServerConnectedEvent` so i need to check that out (but then again, i havent played since 1.16 so maybe that was fixed long ago)

[1]; i don't own multiple accounts and i obviously needed to have different players in different servers to use global chat

unilock committed:
Minor optimization to server join/switch logic

Signed-off-by: unilock <unilock@fennet.rentals>

unilock committed:
Fix "previous_server" name in switch message format

Signed-off-by: unilock <unilock@fennet.rentals>

unilock committed:
Disable cancelling chat messages when global chat passthrough is disabled (which breaks >=1.19.1)

(the config option description didn't account for it anyway - if the cancelling *did* work, duplicate messages would *not* be shown in the sender's chat)

Signed-off-by: unilock <unilock@fennet.rentals>

unilock committed:
Clarify config options + Disable "global chat passthrough" by default

From the perspective of an end-user, it would be odd to see some chat messages duplicated by default

Signed-off-by: unilock <unilock@fennet.rentals>

unilock committed:
Update Gradle + dependencies

Signed-off-by: unilock <unilock@fennet.rentals>

unilock commented:
To break down the most recent commits:
- The config file has been edited to (hopefully) make the options more clear
- The join/switch message sending logic has been slightly optimized, in line with the "exercise for the reader" I mentioned in the OP
- The plugin's dependencies have been updated to their latest versions (except Gradle, since Gradle 8 made some build script syntax changes that I have not tested yet)

More notably:
- Cancelling the chat message sent by the player when "global chat passthrough" is disabled has been removed, as it kicks the player on >=1.19.1 (see #8)
  - The description of the relevant config option did not account for this anyway, as if the cancelling *did* work, the player sending the chat message would *not* see duplicate messages in their chat
- Global chat passthrough is now disabled by default, as I imagine the typical end-user would not expect to see duplicated chat messages by default

I imagine some of these changes could be controversial (especially w.r.t. the default config file), so do let me know what you think.

unilock committed:
Make GLOBAL_CHAT_PASSTHROUGH work like it should

essentially invert all checks for whether it's enabled

Signed-off-by: unilock <unilock@fennet.rentals>

BBaoVanC reviewed (changes requested):
Tested this branch on my proxy, I was using my patch that allows cancelling messages so I had `passthrough = false` in config.

See the following diff for all the changes I made during testing (it should address every review comment I put, if I didn't miss anything):

```diff
diff --git a/gradlew b/gradlew
old mode 100644
new mode 100755
diff --git a/src/main/java/me/confor/velocity/chat/modules/GlobalChat.java b/src/main/java/me/confor/velocity/chat/modules/GlobalChat.java
index 2bab1ae..8160c5f 100644
--- a/src/main/java/me/confor/velocity/chat/modules/GlobalChat.java
+++ b/src/main/java/me/confor/velocity/chat/modules/GlobalChat.java
@@ -4,8 +4,10 @@ import com.velocitypowered.api.event.PostOrder;
 import com.velocitypowered.api.event.Subscribe;
 import com.velocitypowered.api.event.connection.DisconnectEvent;
 import com.velocitypowered.api.event.player.PlayerChatEvent;
+import com.velocitypowered.api.event.player.PlayerChatEvent.ChatResult;
 import com.velocitypowered.api.event.player.ServerConnectedEvent;
 import com.velocitypowered.api.event.proxy.ProxyReloadEvent;
+import com.velocitypowered.api.proxy.Player;
 import com.velocitypowered.api.proxy.ProxyServer;
 import com.velocitypowered.api.proxy.ServerConnection;
 import com.velocitypowered.api.proxy.server.RegisteredServer;
@@ -64,19 +66,13 @@ public class GlobalChat {
             msg = msg.replaceText(config.urlReplacement);
 
         if (config.GLOBAL_CHAT_PASSTHROUGH)
-            sendMessage(msg);
-        else
             sendMessage(msg, currentServer.get().getServer());
+        else
+            sendMessage(msg);
+            event.setResult(ChatResult.denied());
 
         if (config.GLOBAL_CHAT_TO_CONSOLE)
             this.logger.info("GLOBAL: <{}> {}", player, message);
-
-        /*
-        broken >=1.19.1
-
-        if (config.GLOBAL_CHAT_PASSTHROUGH)
-            event.setResult(ChatResult.denied());
-         */
     }
 
     @Subscribe
@@ -87,33 +83,26 @@ public class GlobalChat {
         String player = event.getPlayer().getUsername();
         String server = currentServer.getServerInfo().getName();
 
-        Component msg;
-
         if (previousServer.isPresent()) {
             if (!config.SWITCH_ENABLE)
                 return;
 
-            msg = parseMessage(config.SWITCH_FORMAT, List.of(
+            Component msg = parseMessage(config.SWITCH_FORMAT, List.of(
                 new ChatTemplate("player", player, false),
                 new ChatTemplate("server", server, false),
                 new ChatTemplate("previous_server", previousServer.get().getServerInfo().getName(), false)
             ));
+            sendMessage(msg);
         } else {
             if (!config.JOIN_ENABLE)
                 return;
 
-            msg = parseMessage(config.JOIN_FORMAT, List.of(
+            Component msg = parseMessage(config.JOIN_FORMAT, List.of(
                 new ChatTemplate("player", player, false),
                 new ChatTemplate("server", server, false)
             ));
-
-            if (!config.GLOBAL_CHAT_PASSTHROUGH) {
-                sendMessage(msg, currentServer);
-                return;
-            }
+            sendMessage(msg);
         }
-
-        sendMessage(msg);
     }
 
     @Subscribe
@@ -123,6 +112,7 @@ public class GlobalChat {
 
         Optional<ServerConnection> currentServer = event.getPlayer().getCurrentServer();
         if (currentServer.isEmpty())
+            // TODO: we should find a way to recover and still send a generic disconnect message even if the current server is unknown
             return;
 
         String player = event.getPlayer().getUsername();
@@ -133,10 +123,7 @@ public class GlobalChat {
                 new ChatTemplate("server", server, false)
         ));
 
-        if (config.GLOBAL_CHAT_PASSTHROUGH)
-            sendMessage(msg);
-        else
-            sendMessage(msg, currentServer.get().getServer());
+        sendMessage(msg);
     }
 
     private Component parseMessage(String input, List<ChatTemplate> templates) {
@@ -153,8 +140,8 @@ public class GlobalChat {
     }
 
     private void sendMessage(Component msg) {
-        for (RegisteredServer server : this.server.getAllServers())
-            server.sendMessage(msg);
+        for (Player player : this.server.getAllPlayers())
+            player.sendMessage(msg);
     }
 
     private void sendMessage(Component msg, RegisteredServer currentServer) {
```

unilock commented on gradlew:
I don't even use Windows; not sure how that happened 🤷

unilock commented on src/main/java/me/confor/velocity/chat/modules/GlobalChat.java:
```diff
@@ -66,57 +64,61 @@ public void onPlayerChat(PlayerChatEvent event) {
-        else
             sendMessage(msg);
+        else
+            sendMessage(msg, currentServer.get().getServer());
```
> also `!` was removed from the commented out condition inside the if, presumably on accident?

> Why was this flipped backwards

This is the commit including those changes: https://github.com/confor/velocity-chat/pull/12/commits/dc71853f883cf178834ee2b96493a2a09cdda395

> also am noticing during review that it's confusing how the existing sendMessage overload with a second arg currentServer is to *exclude* that server, maybe function should be named different in the future

I agree. It didn't make sense to me that enabling `GLOBAL_CHAT_PASSTHROUGH` would send messages to all servers *except* the one originally intended to receive the message, hence why I reversed the checks.

unilock commented on src/main/java/me/confor/velocity/chat/modules/GlobalChat.java:
```diff
@@ -66,57 +64,61 @@ public void onPlayerChat(PlayerChatEvent event) {
-            sendMessage(msg);
+            if (!config.GLOBAL_CHAT_PASSTHROUGH) {
+                sendMessage(msg, currentServer);
+                return;
```
I think I made that change because it made sense to me to group chat messages and join messages together under that toggle, as both would be duplicated in the target backend server's chat (by default) with the toggle enabled.

So, with `GLOBAL_CHAT_PASSTHROUGH` enabled, and **without cancelling chat** and disabling join messages on the backend server, players on said backend server would see both their own chat messages duplicated, as well as the join messages of players joining that server.

But yes, a separate toggle sounds like a better idea.

confor replied:
A separate toggle sounds like a good idea.

Since admins might change the join/quit messages, perhaps a good idea is to work with the backend server to hide these possible duplicate messages as unilock suggests. This way the custom join/quit messages are shown network-wide but then the backend server hides its own join/quit vanilla message, so players on that one server don't see the message twice.

unilock commented on src/main/java/me/confor/velocity/chat/modules/GlobalChat.java:
```diff
@@ -66,57 +64,61 @@ public void onPlayerChat(PlayerChatEvent event) {
-    public void onDisconnect(DisconnectEvent event) {
-        if (!config.QUIT_ENABLE)
-            return;
+        Component msg;
```
> is there any advantage here to keeping one single `Component msg` and calling `sendMessage` all the way at the bottom

Realistically, probably not :p

unilock commented on src/main/java/me/confor/velocity/chat/modules/GlobalChat.java:
```diff
@@ -66,57 +64,61 @@ public void onPlayerChat(PlayerChatEvent event) {
+        if (!config.QUIT_ENABLE)
             return;
 
         Optional<ServerConnection> currentServer = event.getPlayer().getCurrentServer();
```
> I think it may be possible for the player to disconnect without the current server being known.

Yes, I think that would happen when a backend server is unreachable. A similar plugin I've worked on, [VelocityDiscord](https://github.com/fooooooooooooooo/VelocityDiscord), implemented something to handle that as well: https://github.com/fooooooooooooooo/VelocityDiscord/commit/64ef162e6afa9e051646c0faa6dceae1f1810ebc

confor replied:
In the past I could reproduce this when connecting through velocity to a server with the wrong client version (minecraft 1.17 to a 1.18 server). Strange things happen to events when you immediately fail to connect.

unilock commented on src/main/java/me/confor/velocity/chat/modules/GlobalChat.java:
```diff
@@ -66,57 +64,61 @@ public void onPlayerChat(PlayerChatEvent event) {
+            }
+        }
+
+        sendMessage(msg);
```
You can't see your own join messages in vanilla Minecraft, but switch messages make sense.

I guess it's better to just show both to any joining player, rather than adhering to vanilla behavior and implementing separate handling just for switch messages.

BBaoVanC replied:
> You can't see your own join messages in vanilla Minecraft

I could when I was testing, at least on 1.19.4

unilock replied:
From my testing, no chat messages are sent to my vanilla 1.19.4 client when first joining a vanilla 1.19.4 server.

<img width="526" alt="Screenshot 2023-05-30 at 8 10 19 AM" src="https://github.com/confor/velocity-chat/assets/11802133/cd1d3b48-8289-4a7f-823d-d9ed21d12eeb">

unilock commented on src/main/resources/config.toml:
```diff
@@ -7,35 +7,44 @@ enable = true
+# if enabled, both the original message and the message
+#   sent globally will appear on the minecraft server
+#   that the message originates from.
+# if disabled, the origin minecraft server will only see
```
I manually justified it, to make it easier to read whenever "word wrap" is turned off. Should I format it differently?
(the extra spaces before the second and third lines are to show that it's a continuation of the first line)

unilock commented on src/main/java/me/confor/velocity/chat/modules/GlobalChat.java:
```diff
@@ -66,57 +64,61 @@ public void onPlayerChat(PlayerChatEvent event) {
+
+        if (config.GLOBAL_CHAT_PASSTHROUGH)
             event.setResult(ChatResult.denied());
+         */
```
> because using other plugins or patches to Velocity you can get it working

UnsignedVelocity can do that: https://modrinth.com/plugin/unsignedvelocity

It's probably at least worth mentioning that such a plugin would be required to use `GLOBAL_CHAT_PASSTHROUGH` on >=1.19.1, maybe in the config file or README.md.

unilock committed:
Changes from https://github.com/confor/velocity-chat/pull/12#pullrequestreview-1447750149

Send messages to all *players* instead of all *servers*, in order to show newly joined players their own join / switch messages

Signed-off-by: unilock <unilock@fennet.rentals>

unilock committed:
Update Adventure

Signed-off-by: unilock <unilock@fennet.rentals>

unilock committed:
More config documentation improvements

Signed-off-by: unilock <unilock@fennet.rentals>

unilock committed:
Clarify parameter name of excluded server in "sendMessage(Component, RegisteredServer)"

Signed-off-by: unilock <unilock@fennet.rentals>

unilock committed:
Allow specifying whether join / leave messages are sent to all servers, or all servers *except* the one being joined / left

Disabled by default, to avoid duplicate messages for those who don't have a mod / plugin that blocks the vanilla join / leave messages on the backend server(s)

Signed-off-by: unilock <unilock@fennet.rentals>

unilock committed:
Update Gradle + Shadow

Signed-off-by: unilock <unilock@fennet.rentals>

unilock committed:
Rename "quit" -> "leave" for vanilla parity

Signed-off-by: unilock <unilock@fennet.rentals>

unilock committed:
Show a message when a player is forcibly disconnected from the proxy

e.g. due to timing out or having an incompatible client

Signed-off-by: unilock <unilock@fennet.rentals>

unilock committed:
Bump config version

Signed-off-by: unilock <unilock@fennet.rentals>

unilock committed:
oops

Signed-off-by: unilock <unilock@fennet.rentals>

confor commented:
i find it absolutely depressing that i _just_ got a notification for this pr.

confor commented:
i'm reading through the changed files. i'm so sorry for taking so long, i genuinely just received an email notification about this pr :(

i'm worried about change from `onConnect(LoginEvent e)` to `onConnect(ServerConnectedEvent e)`, i remember i was able to trigger some issue (join messages not appearing?) by connecting with an outdated mc client.

confor reviewed (approved):
code looks good. i haven't compiled it though.

unilock commented:
> i'm so sorry for taking so long, i genuinely just received an email notification about this pr :(

Don't worry about it; I'm just happy to contribute :)

> i'm worried about change from `onConnect(LoginEvent e)` to `onConnect(ServerConnectedEvent e)`, i remember i was able to trigger some issue (join messages not appearing?) by connecting with an outdated mc client.

It's been working fine for as long as I've been using it, which has been quite a while 🤷‍♀️
